// @smart-webcomponents-angular/breadcrumb/source/modules/smart.maskedtextbox.js

/* Smart UI v9.1.70 (2021-03-15) 
Copyright (c) 2011-2021 jQWidgets. 
License: https://htmlelements.com/license/ */ //

import "../smart.element.js";
import "../smart.button.js";
import "../smart.scrollbar.js";
import "../smart.checkbox.js";
import "../smart.radiobutton.js";
import "../smart.listbox.js";
import "../smart.dropdownlist.js";
import "../smart.combobox.js";
import "../smart.tooltip.js";
import "../smart.numeric.js";
import "../smart.math.js";
import "../smart.complex.js";
import "../smart.textbox.js";
import "../smart.maskedtextbox.js";
import "../smart.passwordtextbox.js";
import "../smart.multilinetextbox.js";
import "../smart.numerictextbox.js";

export class smartTextBox extends Smart.Component {
	get name() {
		return 'smartTextBox';
	}		
}

export class smartMaskedTextBox extends Smart.Component {
	get name() {
		return 'smartMaskedTextBox';
	}		
}

export class smartPasswordTextBox extends Smart.Component {
	get name() {
		return 'smartPasswordTextBox';
	}		
}

export class smartNumericTextBox extends Smart.Component {
	get name() {
		return 'smartNumericTextBox';
	}		
}

export class smartMultilineTextBox extends Smart.Component {
	get name() {
		return 'smartMultilineTextBox';
	}		
}
