// @angular/language-service/ivy/attribute_completions.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/language-service/ivy/attribute_completions", ["require", "exports", "tslib", "@angular/compiler", "typescript", "@angular/language-service/ivy/display_parts", "@angular/language-service/ivy/utils"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getAttributeCompletionSymbol = exports.addAttributeCompletionEntries = exports.buildAttributeCompletionTable = exports.AttributeCompletionKind = void 0;
    var tslib_1 = require("tslib");
    var compiler_1 = require("@angular/compiler");
    var ts = require("typescript");
    var display_parts_1 = require("@angular/language-service/ivy/display_parts");
    var utils_1 = require("@angular/language-service/ivy/utils");
    /**
     * Differentiates different kinds of `AttributeCompletion`s.
     */
    var AttributeCompletionKind;
    (function (AttributeCompletionKind) {
        /**
         * Completion of an attribute from the HTML schema.
         *
         * Attributes often have a corresponding DOM property of the same name.
         */
        AttributeCompletionKind[AttributeCompletionKind["DomAttribute"] = 0] = "DomAttribute";
        /**
         * Completion of a property from the DOM schema.
         *
         * `DomProperty` completions are generated only for properties which don't share their name with
         * an HTML attribute.
         */
        AttributeCompletionKind[AttributeCompletionKind["DomProperty"] = 1] = "DomProperty";
        /**
         * Completion of an attribute that results in a new directive being matched on an element.
         */
        AttributeCompletionKind[AttributeCompletionKind["DirectiveAttribute"] = 2] = "DirectiveAttribute";
        /**
         * Completion of an attribute that results in a new structural directive being matched on an
         * element.
         */
        AttributeCompletionKind[AttributeCompletionKind["StructuralDirectiveAttribute"] = 3] = "StructuralDirectiveAttribute";
        /**
         * Completion of an input from a directive which is either present on the element, or becomes
         * present after the addition of this attribute.
         */
        AttributeCompletionKind[AttributeCompletionKind["DirectiveInput"] = 4] = "DirectiveInput";
        /**
         * Completion of an output from a directive which is either present on the element, or becomes
         * present after the addition of this attribute.
         */
        AttributeCompletionKind[AttributeCompletionKind["DirectiveOutput"] = 5] = "DirectiveOutput";
    })(AttributeCompletionKind = exports.AttributeCompletionKind || (exports.AttributeCompletionKind = {}));
    /**
     * Given an element and its context, produce a `Map` of all possible attribute completions.
     *
     * 3 kinds of attributes are considered for completion, from highest to lowest priority:
     *
     * 1. Inputs/outputs of directives present on the element already.
     * 2. Inputs/outputs of directives that are not present on the element, but which would become
     *    present if such a binding is added.
     * 3. Attributes from the DOM schema for the element.
     *
     * The priority of these options determines which completions are added to the `Map`. If a directive
     * input shares the same name as a DOM attribute, the `Map` will reflect the directive input
     * completion, not the DOM completion for that name.
     */
    function buildAttributeCompletionTable(component, element, checker, includeDomSchemaAttributes) {
        var e_1, _a, e_2, _b, e_3, _c, e_4, _d, e_5, _e, e_6, _f, e_7, _g, e_8, _h;
        var table = new Map();
        // Use the `ElementSymbol` or `TemplateSymbol` to iterate over directives present on the node, and
        // their inputs/outputs. These have the highest priority of completion results.
        var symbol = checker.getSymbolOfNode(element, component);
        var presentDirectives = new Set();
        if (symbol !== null) {
            try {
                // An `ElementSymbol` was available. This means inputs and outputs for directives on the
                // element can be added to the completion table.
                for (var _j = tslib_1.__values(symbol.directives), _k = _j.next(); !_k.done; _k = _j.next()) {
                    var dirSymbol = _k.value;
                    var directive = dirSymbol.tsSymbol.valueDeclaration;
                    if (!ts.isClassDeclaration(directive)) {
                        continue;
                    }
                    presentDirectives.add(directive);
                    var meta = checker.getDirectiveMetadata(directive);
                    if (meta === null) {
                        continue;
                    }
                    try {
                        for (var _l = (e_2 = void 0, tslib_1.__values(meta.inputs)), _m = _l.next(); !_m.done; _m = _l.next()) {
                            var _o = tslib_1.__read(_m.value, 2), classPropertyName = _o[0], propertyName = _o[1];
                            if (table.has(propertyName)) {
                                continue;
                            }
                            table.set(propertyName, {
                                kind: AttributeCompletionKind.DirectiveInput,
                                propertyName: propertyName,
                                directive: dirSymbol,
                                classPropertyName: classPropertyName,
                                twoWayBindingSupported: meta.outputs.hasBindingPropertyName(propertyName + 'Change'),
                            });
                        }
                    }
                    catch (e_2_1) { e_2 = { error: e_2_1 }; }
                    finally {
                        try {
                            if (_m && !_m.done && (_b = _l.return)) _b.call(_l);
                        }
                        finally { if (e_2) throw e_2.error; }
                    }
                    try {
                        for (var _p = (e_3 = void 0, tslib_1.__values(meta.outputs)), _q = _p.next(); !_q.done; _q = _p.next()) {
                            var _r = tslib_1.__read(_q.value, 2), classPropertyName = _r[0], propertyName = _r[1];
                            if (table.has(propertyName)) {
                                continue;
                            }
                            table.set(propertyName, {
                                kind: AttributeCompletionKind.DirectiveOutput,
                                eventName: propertyName,
                                directive: dirSymbol,
                                classPropertyName: classPropertyName,
                            });
                        }
                    }
                    catch (e_3_1) { e_3 = { error: e_3_1 }; }
                    finally {
                        try {
                            if (_q && !_q.done && (_c = _p.return)) _c.call(_p);
                        }
                        finally { if (e_3) throw e_3.error; }
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_k && !_k.done && (_a = _j.return)) _a.call(_j);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        // Next, explore hypothetical directives and determine if the addition of any single attributes
        // can cause the directive to match the element.
        var directivesInScope = checker.getDirectivesInScope(component);
        if (directivesInScope !== null) {
            var elementSelector = utils_1.makeElementSelector(element);
            try {
                for (var directivesInScope_1 = tslib_1.__values(directivesInScope), directivesInScope_1_1 = directivesInScope_1.next(); !directivesInScope_1_1.done; directivesInScope_1_1 = directivesInScope_1.next()) {
                    var dirInScope = directivesInScope_1_1.value;
                    var directive = dirInScope.tsSymbol.valueDeclaration;
                    // Skip directives that are present on the element.
                    if (!ts.isClassDeclaration(directive) || presentDirectives.has(directive)) {
                        continue;
                    }
                    var meta = checker.getDirectiveMetadata(directive);
                    if (meta === null || meta.selector === null) {
                        continue;
                    }
                    if (!meta.isStructural) {
                        // For non-structural directives, the directive's attribute selector(s) are matched against
                        // a hypothetical version of the element with those attributes. A match indicates that
                        // adding that attribute/input/output binding would cause the directive to become present,
                        // meaning that such a binding is a valid completion.
                        var selectors = compiler_1.CssSelector.parse(meta.selector);
                        var matcher = new compiler_1.SelectorMatcher();
                        matcher.addSelectables(selectors);
                        try {
                            for (var selectors_1 = (e_5 = void 0, tslib_1.__values(selectors)), selectors_1_1 = selectors_1.next(); !selectors_1_1.done; selectors_1_1 = selectors_1.next()) {
                                var selector = selectors_1_1.value;
                                try {
                                    for (var _s = (e_6 = void 0, tslib_1.__values(selectorAttributes(selector))), _t = _s.next(); !_t.done; _t = _s.next()) {
                                        var _u = tslib_1.__read(_t.value, 2), attrName = _u[0], attrValue = _u[1];
                                        if (attrValue !== '') {
                                            // This attribute selector requires a value, which is not supported in completion.
                                            continue;
                                        }
                                        if (table.has(attrName)) {
                                            // Skip this attribute as there's already a binding for it.
                                            continue;
                                        }
                                        // Check whether adding this attribute would cause the directive to start matching.
                                        var newElementSelector = elementSelector + ("[" + attrName + "]");
                                        if (!matcher.match(compiler_1.CssSelector.parse(newElementSelector)[0], null)) {
                                            // Nope, move on with our lives.
                                            continue;
                                        }
                                        // Adding this attribute causes a new directive to be matched. Decide how to categorize
                                        // it based on the directive's inputs and outputs.
                                        if (meta.inputs.hasBindingPropertyName(attrName)) {
                                            // This attribute corresponds to an input binding.
                                            table.set(attrName, {
                                                kind: AttributeCompletionKind.DirectiveInput,
                                                directive: dirInScope,
                                                propertyName: attrName,
                                                classPropertyName: meta.inputs.getByBindingPropertyName(attrName)[0].classPropertyName,
                                                twoWayBindingSupported: meta.outputs.hasBindingPropertyName(attrName + 'Change'),
                                            });
                                        }
                                        else if (meta.outputs.hasBindingPropertyName(attrName)) {
                                            // This attribute corresponds to an output binding.
                                            table.set(attrName, {
                                                kind: AttributeCompletionKind.DirectiveOutput,
                                                directive: dirInScope,
                                                eventName: attrName,
                                                classPropertyName: meta.outputs.getByBindingPropertyName(attrName)[0].classPropertyName,
                                            });
                                        }
                                        else {
                                            // This attribute causes a new directive to be matched, but does not also correspond
                                            // to an input or output binding.
                                            table.set(attrName, {
                                                kind: AttributeCompletionKind.DirectiveAttribute,
                                                attribute: attrName,
                                                directive: dirInScope,
                                            });
                                        }
                                    }
                                }
                                catch (e_6_1) { e_6 = { error: e_6_1 }; }
                                finally {
                                    try {
                                        if (_t && !_t.done && (_f = _s.return)) _f.call(_s);
                                    }
                                    finally { if (e_6) throw e_6.error; }
                                }
                            }
                        }
                        catch (e_5_1) { e_5 = { error: e_5_1 }; }
                        finally {
                            try {
                                if (selectors_1_1 && !selectors_1_1.done && (_e = selectors_1.return)) _e.call(selectors_1);
                            }
                            finally { if (e_5) throw e_5.error; }
                        }
                    }
                    else {
                        // Hypothetically matching a structural directive is a litle different than a plain
                        // directive. Use of the '*' structural directive syntactic sugar means that the actual
                        // directive is applied to a plain <ng-template> node, not the existing element with any
                        // other attributes it might already have.
                        // Additionally, more than one attribute/input might need to be present in order for the
                        // directive to match (e.g. `ngFor` has a selector of `[ngFor][ngForOf]`). This gets a
                        // little tricky.
                        var structuralAttributes = getStructuralAttributes(meta);
                        try {
                            for (var structuralAttributes_1 = (e_7 = void 0, tslib_1.__values(structuralAttributes)), structuralAttributes_1_1 = structuralAttributes_1.next(); !structuralAttributes_1_1.done; structuralAttributes_1_1 = structuralAttributes_1.next()) {
                                var attrName = structuralAttributes_1_1.value;
                                table.set(attrName, {
                                    kind: AttributeCompletionKind.StructuralDirectiveAttribute,
                                    attribute: attrName,
                                    directive: dirInScope,
                                });
                            }
                        }
                        catch (e_7_1) { e_7 = { error: e_7_1 }; }
                        finally {
                            try {
                                if (structuralAttributes_1_1 && !structuralAttributes_1_1.done && (_g = structuralAttributes_1.return)) _g.call(structuralAttributes_1);
                            }
                            finally { if (e_7) throw e_7.error; }
                        }
                    }
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (directivesInScope_1_1 && !directivesInScope_1_1.done && (_d = directivesInScope_1.return)) _d.call(directivesInScope_1);
                }
                finally { if (e_4) throw e_4.error; }
            }
        }
        // Finally, add any DOM attributes not already covered by inputs.
        if (element instanceof compiler_1.TmplAstElement && includeDomSchemaAttributes) {
            try {
                for (var _v = tslib_1.__values(checker.getPotentialDomBindings(element.name)), _w = _v.next(); !_w.done; _w = _v.next()) {
                    var _x = _w.value, attribute = _x.attribute, property = _x.property;
                    var isAlsoProperty = attribute === property;
                    if (!table.has(attribute)) {
                        table.set(attribute, {
                            kind: AttributeCompletionKind.DomAttribute,
                            attribute: attribute,
                            isAlsoProperty: isAlsoProperty,
                        });
                    }
                    if (!isAlsoProperty && !table.has(property)) {
                        table.set(property, {
                            kind: AttributeCompletionKind.DomProperty,
                            property: property,
                        });
                    }
                }
            }
            catch (e_8_1) { e_8 = { error: e_8_1 }; }
            finally {
                try {
                    if (_w && !_w.done && (_h = _v.return)) _h.call(_v);
                }
                finally { if (e_8) throw e_8.error; }
            }
        }
        return table;
    }
    exports.buildAttributeCompletionTable = buildAttributeCompletionTable;
    /**
     * Given an `AttributeCompletion`, add any available completions to a `ts.CompletionEntry` array of
     * results.
     *
     * The kind of completions generated depends on whether the current context is an attribute context
     * or not. For example, completing on `<element attr|>` will generate two results: `attribute` and
     * `[attribute]` - either a static attribute can be generated, or a property binding. However,
     * `<element [attr|]>` is not an attribute context, and so only the property completion `attribute`
     * is generated. Note that this completion does not have the `[]` property binding sugar as its
     * implicitly present in a property binding context (we're already completing within an `[attr|]`
     * expression).
     */
    function addAttributeCompletionEntries(entries, completion, isAttributeContext, isElementContext, replacementSpan) {
        switch (completion.kind) {
            case AttributeCompletionKind.DirectiveAttribute: {
                entries.push({
                    kind: display_parts_1.unsafeCastDisplayInfoKindToScriptElementKind(display_parts_1.DisplayInfoKind.DIRECTIVE),
                    name: completion.attribute,
                    sortText: completion.attribute,
                    replacementSpan: replacementSpan,
                });
                break;
            }
            case AttributeCompletionKind.StructuralDirectiveAttribute: {
                // In an element, the completion is offered with a leading '*' to activate the structural
                // directive. Once present, the structural attribute will be parsed as a template and not an
                // element, and the prefix is no longer necessary.
                var prefix = isElementContext ? '*' : '';
                entries.push({
                    kind: display_parts_1.unsafeCastDisplayInfoKindToScriptElementKind(display_parts_1.DisplayInfoKind.DIRECTIVE),
                    name: prefix + completion.attribute,
                    sortText: prefix + completion.attribute,
                    replacementSpan: replacementSpan,
                });
                break;
            }
            case AttributeCompletionKind.DirectiveInput: {
                if (isAttributeContext) {
                    // Offer a completion of a property binding.
                    entries.push({
                        kind: display_parts_1.unsafeCastDisplayInfoKindToScriptElementKind(display_parts_1.DisplayInfoKind.PROPERTY),
                        name: "[" + completion.propertyName + "]",
                        sortText: completion.propertyName,
                        replacementSpan: replacementSpan,
                    });
                    // If the directive supports banana-in-a-box for this input, offer that as well.
                    if (completion.twoWayBindingSupported) {
                        entries.push({
                            kind: display_parts_1.unsafeCastDisplayInfoKindToScriptElementKind(display_parts_1.DisplayInfoKind.PROPERTY),
                            name: "[(" + completion.propertyName + ")]",
                            // This completion should sort after the property binding.
                            sortText: completion.propertyName + '_1',
                            replacementSpan: replacementSpan,
                        });
                    }
                    // Offer a completion of the input binding as an attribute.
                    entries.push({
                        kind: display_parts_1.unsafeCastDisplayInfoKindToScriptElementKind(display_parts_1.DisplayInfoKind.ATTRIBUTE),
                        name: completion.propertyName,
                        // This completion should sort after both property binding options (one-way and two-way).
                        sortText: completion.propertyName + '_2',
                        replacementSpan: replacementSpan,
                    });
                }
                else {
                    entries.push({
                        kind: display_parts_1.unsafeCastDisplayInfoKindToScriptElementKind(display_parts_1.DisplayInfoKind.PROPERTY),
                        name: completion.propertyName,
                        sortText: completion.propertyName,
                        replacementSpan: replacementSpan,
                    });
                }
                break;
            }
            case AttributeCompletionKind.DirectiveOutput: {
                if (isAttributeContext) {
                    entries.push({
                        kind: display_parts_1.unsafeCastDisplayInfoKindToScriptElementKind(display_parts_1.DisplayInfoKind.EVENT),
                        name: "(" + completion.eventName + ")",
                        sortText: completion.eventName,
                        replacementSpan: replacementSpan,
                    });
                }
                else {
                    entries.push({
                        kind: display_parts_1.unsafeCastDisplayInfoKindToScriptElementKind(display_parts_1.DisplayInfoKind.EVENT),
                        name: completion.eventName,
                        sortText: completion.eventName,
                        replacementSpan: replacementSpan,
                    });
                }
                break;
            }
            case AttributeCompletionKind.DomAttribute: {
                if (isAttributeContext) {
                    // Offer a completion of an attribute binding.
                    entries.push({
                        kind: display_parts_1.unsafeCastDisplayInfoKindToScriptElementKind(display_parts_1.DisplayInfoKind.ATTRIBUTE),
                        name: completion.attribute,
                        sortText: completion.attribute,
                        replacementSpan: replacementSpan,
                    });
                    if (completion.isAlsoProperty) {
                        // Offer a completion of a property binding to the DOM property.
                        entries.push({
                            kind: display_parts_1.unsafeCastDisplayInfoKindToScriptElementKind(display_parts_1.DisplayInfoKind.PROPERTY),
                            name: "[" + completion.attribute + "]",
                            // In the case of DOM attributes, the property binding should sort after the attribute
                            // binding.
                            sortText: completion.attribute + '_1',
                            replacementSpan: replacementSpan,
                        });
                    }
                }
                else if (completion.isAlsoProperty) {
                    entries.push({
                        kind: display_parts_1.unsafeCastDisplayInfoKindToScriptElementKind(display_parts_1.DisplayInfoKind.PROPERTY),
                        name: completion.attribute,
                        sortText: completion.attribute,
                        replacementSpan: replacementSpan,
                    });
                }
                break;
            }
            case AttributeCompletionKind.DomProperty: {
                if (!isAttributeContext) {
                    entries.push({
                        kind: display_parts_1.unsafeCastDisplayInfoKindToScriptElementKind(display_parts_1.DisplayInfoKind.PROPERTY),
                        name: completion.property,
                        sortText: completion.property,
                        replacementSpan: replacementSpan,
                    });
                }
            }
        }
    }
    exports.addAttributeCompletionEntries = addAttributeCompletionEntries;
    function getAttributeCompletionSymbol(completion, checker) {
        var _a;
        switch (completion.kind) {
            case AttributeCompletionKind.DomAttribute:
            case AttributeCompletionKind.DomProperty:
                return null;
            case AttributeCompletionKind.DirectiveAttribute:
            case AttributeCompletionKind.StructuralDirectiveAttribute:
                return completion.directive.tsSymbol;
            case AttributeCompletionKind.DirectiveInput:
            case AttributeCompletionKind.DirectiveOutput:
                return (_a = checker.getDeclaredTypeOfSymbol(completion.directive.tsSymbol)
                    .getProperty(completion.classPropertyName)) !== null && _a !== void 0 ? _a : null;
        }
    }
    exports.getAttributeCompletionSymbol = getAttributeCompletionSymbol;
    /**
     * Iterates over `CssSelector` attributes, which are internally represented in a zipped array style
     * which is not conducive to straightforward iteration.
     */
    function selectorAttributes(selector) {
        var i;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    i = 0;
                    _a.label = 1;
                case 1:
                    if (!(i < selector.attrs.length)) return [3 /*break*/, 4];
                    return [4 /*yield*/, [selector.attrs[0], selector.attrs[1]]];
                case 2:
                    _a.sent();
                    _a.label = 3;
                case 3:
                    i += 2;
                    return [3 /*break*/, 1];
                case 4: return [2 /*return*/];
            }
        });
    }
    function getStructuralAttributes(meta) {
        var e_9, _a;
        if (meta.selector === null) {
            return [];
        }
        var structuralAttributes = [];
        var selectors = compiler_1.CssSelector.parse(meta.selector);
        var _loop_1 = function (selector) {
            if (selector.element !== null && selector.element !== 'ng-template') {
                return "continue";
            }
            // Every attribute of this selector must be name-only - no required values.
            var attributeSelectors = Array.from(selectorAttributes(selector));
            if (!attributeSelectors.every(function (_a) {
                var _b = tslib_1.__read(_a, 2), _ = _b[0], attrValue = _b[1];
                return attrValue === '';
            })) {
                return "continue";
            }
            // Get every named selector.
            var attributes = attributeSelectors.map(function (_a) {
                var _b = tslib_1.__read(_a, 2), attrName = _b[0], _ = _b[1];
                return attrName;
            });
            // Find the shortest attribute. This is the structural directive "base", and all potential
            // input bindings must begin with the base. E.g. in `*ngFor="let a of b"`, `ngFor` is the
            // base attribute, and the `of` binding key corresponds to an input of `ngForOf`.
            var baseAttr = attributes.reduce(function (prev, curr) { return prev === null || curr.length < prev.length ? curr : prev; }, null);
            if (baseAttr === null) {
                return "continue";
            }
            // Validate that the attributes are compatible with use as a structural directive.
            var isValid = function (attr) {
                // The base attribute is valid by default.
                if (attr === baseAttr) {
                    return true;
                }
                // Non-base attributes must all be prefixed with the base attribute.
                if (!attr.startsWith(baseAttr)) {
                    return false;
                }
                // Non-base attributes must also correspond to directive inputs.
                if (!meta.inputs.hasBindingPropertyName(attr)) {
                    return false;
                }
                // This attribute is compatible.
                return true;
            };
            if (!attributes.every(isValid)) {
                return "continue";
            }
            // This attribute is valid as a structural attribute for this directive.
            structuralAttributes.push(baseAttr);
        };
        try {
            for (var selectors_2 = tslib_1.__values(selectors), selectors_2_1 = selectors_2.next(); !selectors_2_1.done; selectors_2_1 = selectors_2.next()) {
                var selector = selectors_2_1.value;
                _loop_1(selector);
            }
        }
        catch (e_9_1) { e_9 = { error: e_9_1 }; }
        finally {
            try {
                if (selectors_2_1 && !selectors_2_1.done && (_a = selectors_2.return)) _a.call(selectors_2);
            }
            finally { if (e_9) throw e_9.error; }
        }
        return structuralAttributes;
    }
});
