// @angular/language-service/ivy/display_parts.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/language-service/ivy/display_parts", ["require", "exports", "tslib", "@angular/compiler-cli/src/ngtsc/typecheck/api", "typescript"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getTsSymbolDisplayInfo = exports.getDirectiveDisplayInfo = exports.unsafeCastDisplayInfoKindToScriptElementKind = exports.createDisplayParts = exports.getSymbolDisplayInfo = exports.DisplayInfoKind = exports.SYMBOL_TEXT = exports.SYMBOL_SPACE = exports.SYMBOL_PUNC = exports.SYMBOL_INTERFACE = exports.ALIAS_NAME = void 0;
    var tslib_1 = require("tslib");
    var api_1 = require("@angular/compiler-cli/src/ngtsc/typecheck/api");
    var ts = require("typescript");
    // Reverse mappings of enum would generate strings
    exports.ALIAS_NAME = ts.SymbolDisplayPartKind[ts.SymbolDisplayPartKind.aliasName];
    exports.SYMBOL_INTERFACE = ts.SymbolDisplayPartKind[ts.SymbolDisplayPartKind.interfaceName];
    exports.SYMBOL_PUNC = ts.SymbolDisplayPartKind[ts.SymbolDisplayPartKind.punctuation];
    exports.SYMBOL_SPACE = ts.SymbolDisplayPartKind[ts.SymbolDisplayPartKind.space];
    exports.SYMBOL_TEXT = ts.SymbolDisplayPartKind[ts.SymbolDisplayPartKind.text];
    /**
     * Label for various kinds of Angular entities for TS display info.
     */
    var DisplayInfoKind;
    (function (DisplayInfoKind) {
        DisplayInfoKind["ATTRIBUTE"] = "attribute";
        DisplayInfoKind["COMPONENT"] = "component";
        DisplayInfoKind["DIRECTIVE"] = "directive";
        DisplayInfoKind["EVENT"] = "event";
        DisplayInfoKind["REFERENCE"] = "reference";
        DisplayInfoKind["ELEMENT"] = "element";
        DisplayInfoKind["VARIABLE"] = "variable";
        DisplayInfoKind["PIPE"] = "pipe";
        DisplayInfoKind["PROPERTY"] = "property";
        DisplayInfoKind["METHOD"] = "method";
        DisplayInfoKind["TEMPLATE"] = "template";
    })(DisplayInfoKind = exports.DisplayInfoKind || (exports.DisplayInfoKind = {}));
    function getSymbolDisplayInfo(tsLS, typeChecker, symbol) {
        var kind;
        if (symbol.kind === api_1.SymbolKind.Reference) {
            kind = DisplayInfoKind.REFERENCE;
        }
        else if (symbol.kind === api_1.SymbolKind.Variable) {
            kind = DisplayInfoKind.VARIABLE;
        }
        else {
            throw new Error("AssertionError: unexpected symbol kind " + api_1.SymbolKind[symbol.kind]);
        }
        var displayParts = createDisplayParts(symbol.declaration.name, kind, /* containerName */ undefined, typeChecker.typeToString(symbol.tsType));
        var documentation = symbol.kind === api_1.SymbolKind.Reference ?
            getDocumentationFromTypeDefAtLocation(tsLS, symbol.targetLocation) :
            getDocumentationFromTypeDefAtLocation(tsLS, symbol.initializerLocation);
        return {
            kind: kind,
            displayParts: displayParts,
            documentation: documentation,
        };
    }
    exports.getSymbolDisplayInfo = getSymbolDisplayInfo;
    /**
     * Construct a compound `ts.SymbolDisplayPart[]` which incorporates the container and type of a
     * target declaration.
     * @param name Name of the target
     * @param kind component, directive, pipe, etc.
     * @param containerName either the Symbol's container or the NgModule that contains the directive
     * @param type user-friendly name of the type
     * @param documentation docstring or comment
     */
    function createDisplayParts(name, kind, containerName, type) {
        var containerDisplayParts = containerName !== undefined ?
            [
                { text: containerName, kind: exports.SYMBOL_INTERFACE },
                { text: '.', kind: exports.SYMBOL_PUNC },
            ] :
            [];
        var typeDisplayParts = type !== undefined ?
            [
                { text: ':', kind: exports.SYMBOL_PUNC },
                { text: ' ', kind: exports.SYMBOL_SPACE },
                { text: type, kind: exports.SYMBOL_INTERFACE },
            ] :
            [];
        return tslib_1.__spread([
            { text: '(', kind: exports.SYMBOL_PUNC },
            { text: kind, kind: exports.SYMBOL_TEXT },
            { text: ')', kind: exports.SYMBOL_PUNC },
            { text: ' ', kind: exports.SYMBOL_SPACE }
        ], containerDisplayParts, [
            { text: name, kind: exports.SYMBOL_INTERFACE }
        ], typeDisplayParts);
    }
    exports.createDisplayParts = createDisplayParts;
    /**
     * Convert a `SymbolDisplayInfoKind` to a `ts.ScriptElementKind` type, allowing it to pass through
     * TypeScript APIs.
     *
     * In practice, this is an "illegal" type cast. Since `ts.ScriptElementKind` is a string, this is
     * safe to do if TypeScript only uses the value in a string context. Consumers of this conversion
     * function are responsible for ensuring this is the case.
     */
    function unsafeCastDisplayInfoKindToScriptElementKind(kind) {
        return kind;
    }
    exports.unsafeCastDisplayInfoKindToScriptElementKind = unsafeCastDisplayInfoKindToScriptElementKind;
    function getDocumentationFromTypeDefAtLocation(tsLS, shimLocation) {
        var _a;
        var typeDefs = tsLS.getTypeDefinitionAtPosition(shimLocation.shimPath, shimLocation.positionInShimFile);
        if (typeDefs === undefined || typeDefs.length === 0) {
            return undefined;
        }
        return (_a = tsLS.getQuickInfoAtPosition(typeDefs[0].fileName, typeDefs[0].textSpan.start)) === null || _a === void 0 ? void 0 : _a.documentation;
    }
    function getDirectiveDisplayInfo(tsLS, dir) {
        var _a, _b;
        var kind = dir.isComponent ? DisplayInfoKind.COMPONENT : DisplayInfoKind.DIRECTIVE;
        var decl = dir.tsSymbol.declarations.find(ts.isClassDeclaration);
        if (decl === undefined || decl.name === undefined) {
            return { kind: kind, displayParts: [], documentation: [] };
        }
        var res = tsLS.getQuickInfoAtPosition(decl.getSourceFile().fileName, decl.name.getStart());
        if (res === undefined) {
            return { kind: kind, displayParts: [], documentation: [] };
        }
        var displayParts = createDisplayParts(dir.tsSymbol.name, kind, (_b = (_a = dir.ngModule) === null || _a === void 0 ? void 0 : _a.name) === null || _b === void 0 ? void 0 : _b.text, undefined);
        return {
            kind: kind,
            displayParts: displayParts,
            documentation: res.documentation,
        };
    }
    exports.getDirectiveDisplayInfo = getDirectiveDisplayInfo;
    function getTsSymbolDisplayInfo(tsLS, checker, symbol, kind, ownerName) {
        var decl = symbol.valueDeclaration;
        if (decl === undefined || (!ts.isPropertyDeclaration(decl) && !ts.isMethodDeclaration(decl)) ||
            !ts.isIdentifier(decl.name)) {
            return null;
        }
        var res = tsLS.getQuickInfoAtPosition(decl.getSourceFile().fileName, decl.name.getStart());
        if (res === undefined) {
            return { kind: kind, displayParts: [], documentation: [] };
        }
        var type = checker.getDeclaredTypeOfSymbol(symbol);
        var typeString = checker.typeToString(type);
        var displayParts = createDisplayParts(symbol.name, kind, ownerName !== null && ownerName !== void 0 ? ownerName : undefined, typeString);
        return {
            kind: kind,
            displayParts: displayParts,
            documentation: res.documentation,
        };
    }
    exports.getTsSymbolDisplayInfo = getTsSymbolDisplayInfo;
});
