// @angular/language-service/ivy/quick_info.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/language-service/ivy/quick_info", ["require", "exports", "tslib", "@angular/compiler", "@angular/compiler-cli/src/ngtsc/typecheck/api", "typescript", "@angular/language-service/ivy/display_parts", "@angular/language-service/ivy/utils"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.createQuickInfo = exports.QuickInfoBuilder = void 0;
    var tslib_1 = require("tslib");
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    var compiler_1 = require("@angular/compiler");
    var api_1 = require("@angular/compiler-cli/src/ngtsc/typecheck/api");
    var ts = require("typescript");
    var display_parts_1 = require("@angular/language-service/ivy/display_parts");
    var utils_1 = require("@angular/language-service/ivy/utils");
    var QuickInfoBuilder = /** @class */ (function () {
        function QuickInfoBuilder(tsLS, compiler, component, node) {
            this.tsLS = tsLS;
            this.compiler = compiler;
            this.component = component;
            this.node = node;
            this.typeChecker = this.compiler.getNextProgram().getTypeChecker();
        }
        QuickInfoBuilder.prototype.get = function () {
            var symbol = this.compiler.getTemplateTypeChecker().getSymbolOfNode(this.node, this.component);
            if (symbol === null) {
                return isDollarAny(this.node) ? createDollarAnyQuickInfo(this.node) : undefined;
            }
            return this.getQuickInfoForSymbol(symbol);
        };
        QuickInfoBuilder.prototype.getQuickInfoForSymbol = function (symbol) {
            switch (symbol.kind) {
                case api_1.SymbolKind.Input:
                case api_1.SymbolKind.Output:
                    return this.getQuickInfoForBindingSymbol(symbol);
                case api_1.SymbolKind.Template:
                    return createNgTemplateQuickInfo(this.node);
                case api_1.SymbolKind.Element:
                    return this.getQuickInfoForElementSymbol(symbol);
                case api_1.SymbolKind.Variable:
                    return this.getQuickInfoForVariableSymbol(symbol);
                case api_1.SymbolKind.Reference:
                    return this.getQuickInfoForReferenceSymbol(symbol);
                case api_1.SymbolKind.DomBinding:
                    return this.getQuickInfoForDomBinding(symbol);
                case api_1.SymbolKind.Directive:
                    return this.getQuickInfoAtShimLocation(symbol.shimLocation);
                case api_1.SymbolKind.Pipe:
                    return this.getQuickInfoForPipeSymbol(symbol);
                case api_1.SymbolKind.Expression:
                    return this.getQuickInfoAtShimLocation(symbol.shimLocation);
            }
        };
        QuickInfoBuilder.prototype.getQuickInfoForBindingSymbol = function (symbol) {
            if (symbol.bindings.length === 0) {
                return undefined;
            }
            var kind = symbol.kind === api_1.SymbolKind.Input ? display_parts_1.DisplayInfoKind.PROPERTY : display_parts_1.DisplayInfoKind.EVENT;
            var quickInfo = this.getQuickInfoAtShimLocation(symbol.bindings[0].shimLocation);
            return quickInfo === undefined ? undefined : updateQuickInfoKind(quickInfo, kind);
        };
        QuickInfoBuilder.prototype.getQuickInfoForElementSymbol = function (symbol) {
            var templateNode = symbol.templateNode;
            var matches = utils_1.getDirectiveMatchesForElementTag(templateNode, symbol.directives);
            if (matches.size > 0) {
                return this.getQuickInfoForDirectiveSymbol(matches.values().next().value, templateNode);
            }
            return createQuickInfo(templateNode.name, display_parts_1.DisplayInfoKind.ELEMENT, utils_1.getTextSpanOfNode(templateNode), undefined /* containerName */, this.typeChecker.typeToString(symbol.tsType));
        };
        QuickInfoBuilder.prototype.getQuickInfoForVariableSymbol = function (symbol) {
            var documentation = this.getDocumentationFromTypeDefAtLocation(symbol.initializerLocation);
            return createQuickInfo(symbol.declaration.name, display_parts_1.DisplayInfoKind.VARIABLE, utils_1.getTextSpanOfNode(this.node), undefined /* containerName */, this.typeChecker.typeToString(symbol.tsType), documentation);
        };
        QuickInfoBuilder.prototype.getQuickInfoForReferenceSymbol = function (symbol) {
            var documentation = this.getDocumentationFromTypeDefAtLocation(symbol.targetLocation);
            return createQuickInfo(symbol.declaration.name, display_parts_1.DisplayInfoKind.REFERENCE, utils_1.getTextSpanOfNode(this.node), undefined /* containerName */, this.typeChecker.typeToString(symbol.tsType), documentation);
        };
        QuickInfoBuilder.prototype.getQuickInfoForPipeSymbol = function (symbol) {
            if (symbol.tsSymbol !== null) {
                var quickInfo = this.getQuickInfoAtShimLocation(symbol.shimLocation);
                return quickInfo === undefined ? undefined :
                    updateQuickInfoKind(quickInfo, display_parts_1.DisplayInfoKind.PIPE);
            }
            else {
                return createQuickInfo(this.typeChecker.typeToString(symbol.classSymbol.tsType), display_parts_1.DisplayInfoKind.PIPE, utils_1.getTextSpanOfNode(this.node));
            }
        };
        QuickInfoBuilder.prototype.getQuickInfoForDomBinding = function (symbol) {
            if (!(this.node instanceof compiler_1.TmplAstTextAttribute) &&
                !(this.node instanceof compiler_1.TmplAstBoundAttribute)) {
                return undefined;
            }
            var directives = utils_1.getDirectiveMatchesForAttribute(this.node.name, symbol.host.templateNode, symbol.host.directives);
            if (directives.size === 0) {
                return undefined;
            }
            return this.getQuickInfoForDirectiveSymbol(directives.values().next().value);
        };
        QuickInfoBuilder.prototype.getQuickInfoForDirectiveSymbol = function (dir, node) {
            if (node === void 0) { node = this.node; }
            var kind = dir.isComponent ? display_parts_1.DisplayInfoKind.COMPONENT : display_parts_1.DisplayInfoKind.DIRECTIVE;
            var documentation = this.getDocumentationFromTypeDefAtLocation(dir.shimLocation);
            var containerName;
            if (ts.isClassDeclaration(dir.tsSymbol.valueDeclaration) && dir.ngModule !== null) {
                containerName = dir.ngModule.name.getText();
            }
            return createQuickInfo(this.typeChecker.typeToString(dir.tsType), kind, utils_1.getTextSpanOfNode(this.node), containerName, undefined, documentation);
        };
        QuickInfoBuilder.prototype.getDocumentationFromTypeDefAtLocation = function (shimLocation) {
            var _a;
            var typeDefs = this.tsLS.getTypeDefinitionAtPosition(shimLocation.shimPath, shimLocation.positionInShimFile);
            if (typeDefs === undefined || typeDefs.length === 0) {
                return undefined;
            }
            return (_a = this.tsLS.getQuickInfoAtPosition(typeDefs[0].fileName, typeDefs[0].textSpan.start)) === null || _a === void 0 ? void 0 : _a.documentation;
        };
        QuickInfoBuilder.prototype.getQuickInfoAtShimLocation = function (location) {
            var quickInfo = this.tsLS.getQuickInfoAtPosition(location.shimPath, location.positionInShimFile);
            if (quickInfo === undefined || quickInfo.displayParts === undefined) {
                return quickInfo;
            }
            quickInfo.displayParts = utils_1.filterAliasImports(quickInfo.displayParts);
            var textSpan = utils_1.getTextSpanOfNode(this.node);
            return tslib_1.__assign(tslib_1.__assign({}, quickInfo), { textSpan: textSpan });
        };
        return QuickInfoBuilder;
    }());
    exports.QuickInfoBuilder = QuickInfoBuilder;
    function updateQuickInfoKind(quickInfo, kind) {
        if (quickInfo.displayParts === undefined) {
            return quickInfo;
        }
        var startsWithKind = quickInfo.displayParts.length >= 3 &&
            displayPartsEqual(quickInfo.displayParts[0], { text: '(', kind: display_parts_1.SYMBOL_PUNC }) &&
            quickInfo.displayParts[1].kind === display_parts_1.SYMBOL_TEXT &&
            displayPartsEqual(quickInfo.displayParts[2], { text: ')', kind: display_parts_1.SYMBOL_PUNC });
        if (startsWithKind) {
            quickInfo.displayParts[1].text = kind;
        }
        else {
            quickInfo.displayParts = tslib_1.__spread([
                { text: '(', kind: display_parts_1.SYMBOL_PUNC },
                { text: kind, kind: display_parts_1.SYMBOL_TEXT },
                { text: ')', kind: display_parts_1.SYMBOL_PUNC },
                { text: ' ', kind: display_parts_1.SYMBOL_SPACE }
            ], quickInfo.displayParts);
        }
        return quickInfo;
    }
    function displayPartsEqual(a, b) {
        return a.text === b.text && a.kind === b.kind;
    }
    function isDollarAny(node) {
        return node instanceof compiler_1.MethodCall && node.receiver instanceof compiler_1.ImplicitReceiver &&
            !(node.receiver instanceof compiler_1.ThisReceiver) && node.name === '$any' && node.args.length === 1;
    }
    function createDollarAnyQuickInfo(node) {
        return createQuickInfo('$any', display_parts_1.DisplayInfoKind.METHOD, utils_1.getTextSpanOfNode(node), 
        /** containerName */ undefined, 'any', [{
                kind: display_parts_1.SYMBOL_TEXT,
                text: 'function to cast an expression to the `any` type',
            }]);
    }
    // TODO(atscott): Create special `ts.QuickInfo` for `ng-template` and `ng-container` as well.
    function createNgTemplateQuickInfo(node) {
        return createQuickInfo('ng-template', display_parts_1.DisplayInfoKind.TEMPLATE, utils_1.getTextSpanOfNode(node), 
        /** containerName */ undefined, 
        /** type */ undefined, [{
                kind: display_parts_1.SYMBOL_TEXT,
                text: 'The `<ng-template>` is an Angular element for rendering HTML. It is never displayed directly.',
            }]);
    }
    /**
     * Construct a QuickInfo object taking into account its container and type.
     * @param name Name of the QuickInfo target
     * @param kind component, directive, pipe, etc.
     * @param textSpan span of the target
     * @param containerName either the Symbol's container or the NgModule that contains the directive
     * @param type user-friendly name of the type
     * @param documentation docstring or comment
     */
    function createQuickInfo(name, kind, textSpan, containerName, type, documentation) {
        var displayParts = display_parts_1.createDisplayParts(name, kind, containerName, type);
        return {
            kind: display_parts_1.unsafeCastDisplayInfoKindToScriptElementKind(kind),
            kindModifiers: ts.ScriptElementKindModifier.none,
            textSpan: textSpan,
            displayParts: displayParts,
            documentation: documentation,
        };
    }
    exports.createQuickInfo = createQuickInfo;
});
